### PR TITLE
Bugfix/multipart

### LINKED ISSUE #441: Logbook HttpRequest/HttpResponse TCK
## Detailed Description

Define and document the contract of `HttpRequest`/`HttpResponse`'s `withBody`, `withoutBody` and `getBody` opoerations and provide a [*"Technology Compatibility Kit"*](https://en.wikipedia.org/wiki/Technology_Compatibility_Kit) for implementations to rely on.

### Possible states and transitions 

[image](http://bit.ly/2EjCW9s)

### Ideas

Buffering is an internal state which we don't expose right now. For a TCK it would be helpful to verify the desired effect (not buffering unless needed) was actually applied. Maybe that can be achieved by delegating back to the implementation-specific part of the test suite.

- [ ] State pattern within request/response implementations?
- [ ] TCK based on one or more `Strategy` implementations which verify that all transitions work as expected (including negative tests)
- [ ] Is it even possible to have `withBody` and `withoutBody` return new instances with modified state?

## Your Environment

* Version used: 2.0.0-SNAPSHOT

### LINKED ISSUE #637: Can not handle multipart/form-data
Hello, I have problem with logbook. When I create endpoint where I am able to upload the file everything works correctly. After adding logbook dependency I am not able to work with the uploaded file. 

## Steps to Reproduce
For better understanding I have created minimal example with bug: https://github.com/radovans/logbook-problem

when I call the endpoint with postman:
`curl -X POST \
  http://localhost:8888/upload/file \
  -H 'Accept: */*' \
  -H 'Accept-Encoding: gzip, deflate' \
  -H 'Connection: keep-alive' \
  -H 'Content-Length: 15001' \
  -H 'Content-Type: multipart/form-data; boundary=--------------------------606195755734019197267982' \
  -H 'Host: localhost:8888' \
  -H 'User-Agent: PostmanRuntime/7.19.0' \
  -H 'content-type: multipart/form-data; boundary=----WebKitFormBoundary7MA4YWxkTrZu0gW' \
  -F file=@resume.pdf`

postman

I get error

`{
    "timestamp": "2019-11-13T14:23:42.887+0000",
    "status": 400,
    "error": "Bad Request",
    "message": "Required request part 'file' is not present",
    "path": "/upload/file"
}`

so I am not able to handle the file and store it for example. When i remove all logbook dependencies I have no problem with that.

### PR DESCRIPTION
**Pull request by whiskeysierra**
## Description

Implemented state pattern for request/response implementations.

## Motivation and Context

Fixes #637 and fixes #441 

We now have a consistent pattern to implement the internal states of requests and responses. I started it to fix #637, but I don't have a good regression test for that yet.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: logbook-okhttp2/src/main/java/org/zalando/logbook/okhttp2/LocalRequest.java [hunk 1]
```diff
@@ -13,20 +16,126 @@
...
+
+    private final class Passing extends AbstractState {
 
-    LocalRequest(final Request request) {
-        this.request = request;
     }
 
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
+
+    private final Request request;
```
Comment (whiskeysierra) [inline logbook-okhttp2/src/main/java/org/zalando/logbook/okhttp2/LocalRequest.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-okhttp2/src/main/java/org/zalando/logbook/okhttp2/RemoteResponse.java [hunk 2]
```diff
@@ -12,20 +14,124 @@
...
+
+    }
+
+    private final class Passing extends AbstractState {
+
+    }
+
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
+
+    private final Response response;
```
Comment (whiskeysierra) [inline logbook-okhttp2/src/main/java/org/zalando/logbook/okhttp2/RemoteResponse.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-httpclient/src/main/java/org/zalando/logbook/httpclient/RemoteResponse.java [hunk 3]
```diff
@@ -14,23 +15,113 @@
...
+            return new Buffering(body);
+        }
+
+    }
+
+    private static final class Passing implements State {
+
+    }
+
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```
Comment (whiskeysierra) [inline logbook-httpclient/src/main/java/org/zalando/logbook/httpclient/RemoteResponse.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-okhttp/src/main/java/org/zalando/logbook/okhttp/RemoteResponse.java [hunk 4]
```diff
@@ -1,33 +1,137 @@
...
+            return new Buffering(response, body);
+        }
+
+    }
+
+    private final class Passing extends AbstractState {
+
+    }
+
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```
Comment (whiskeysierra) [inline logbook-okhttp/src/main/java/org/zalando/logbook/okhttp/RemoteResponse.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/LocalRequest.java [hunk 5]
```diff
@@ -1,28 +1,101 @@
...
+        @Override
+        public State with() {
+            return new Buffering(stream);
+        }
 
-    public LocalRequest(final ClientRequestContext context) {
-        this.context = context;
     }
 
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```
Comment (whiskeysierra) [inline logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/LocalRequest.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-okhttp/src/main/java/org/zalando/logbook/okhttp/RemoteResponse.java [hunk 6]
```diff
@@ -1,33 +1,137 @@
...
+
+    private final Response response;
```
Comment (whiskeysierra) [inline logbook-okhttp/src/main/java/org/zalando/logbook/okhttp/RemoteResponse.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-servlet/src/main/java/org/zalando/logbook/servlet/RemoteRequest.java [hunk 7]
```diff
@@ -1,36 +1,192 @@
...
+            return new Buffering(body, getStream());
+        }
+
+    }
+
+    private static final class Passing implements State {
+
+    }
+
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```
Comment (whiskeysierra) [inline logbook-servlet/src/main/java/org/zalando/logbook/servlet/RemoteRequest.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-servlet/src/main/java/org/zalando/logbook/servlet/LocalResponse.java [hunk 8]
```diff
@@ -19,17 +21,135 @@
...
+        }
+
+        default PrintWriter getWriter(
+                final ServletResponse response,
+                final Supplier<Charset> charset) throws IOException {
+
+            return response.getWriter();
+        }
+
+        default void flush() throws IOException {
```
Comment (whiskeysierra) [inline logbook-servlet/src/main/java/org/zalando/logbook/servlet/LocalResponse.java]: Codacy Issue found: [Document empty method body](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-servlet/src/main/java/org/zalando/logbook/servlet/LocalResponse.java [hunk 9]
```diff
@@ -19,17 +21,135 @@
...
+        }
+
+        @Override
+        public State with() {
+            return new Buffering(getTee());
+        }
+
+    }
+
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```
Comment (whiskeysierra) [inline logbook-servlet/src/main/java/org/zalando/logbook/servlet/LocalResponse.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/RemoteResponse.java [hunk 10]
```diff
@@ -11,16 +12,94 @@
...
+        @Override
+        public State with() {
+            return new Buffering(body);
+        }
+
+    }
+
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
+
+    private final ClientResponseContext context;
```
Comment (whiskeysierra) [inline logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/RemoteResponse.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/LocalResponse.java [hunk 11]
```diff
@@ -9,18 +10,89 @@
...
+            return new Buffering(stream);
+        }
 
-    public LocalResponse(final ContainerResponseContext context) {
-        this.context = context;
     }
 
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
+
+    private final ContainerResponseContext context;
```
Comment (whiskeysierra) [inline logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/LocalResponse.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/RemoteRequest.java [hunk 12]
```diff
@@ -11,18 +12,95 @@
...
+        @Override
+        public State with() {
+            return new Buffering(body);
+        }
 
-    public RemoteRequest(final ContainerRequestContext context) {
-        this.context = context;
     }
 
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```
Comment (whiskeysierra) [inline logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/RemoteRequest.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-okhttp2/src/main/java/org/zalando/logbook/okhttp2/LocalRequest.java [hunk 13]
```diff
@@ -3,7 +3,10 @@
 import com.squareup.okhttp.MediaType;
 import com.squareup.okhttp.Request;
 import com.squareup.okhttp.RequestBody;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
 import okio.Buffer;
+import org.zalando.fauxpas.FauxPas;
```
Comment (whiskeysierra) [inline logbook-okhttp2/src/main/java/org/zalando/logbook/okhttp2/LocalRequest.java]: Codacy Issue found: [Avoid unused imports such as 'org.zalando.fauxpas.FauxPas'](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-httpclient/src/main/java/org/zalando/logbook/httpclient/RemoteResponse.java [hunk 14]
```diff
@@ -14,23 +15,113 @@
...
+    private final HttpResponse response;
```
Comment (whiskeysierra) [inline logbook-httpclient/src/main/java/org/zalando/logbook/httpclient/RemoteResponse.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-okhttp2/src/main/java/org/zalando/logbook/okhttp2/RemoteResponse.java [hunk 15]
```diff
@@ -12,20 +14,124 @@
...
+            return new Buffering(response, body);
+        }
+
+    }
+
+    private final class Passing extends AbstractState {
+
+    }
+
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```
Comment (whiskeysierra) [inline logbook-okhttp2/src/main/java/org/zalando/logbook/okhttp2/RemoteResponse.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/LocalRequest.java [hunk 16]
```diff
@@ -1,28 +1,101 @@
...
+
+    private final ClientRequestContext context;
```
Comment (whiskeysierra) [inline logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/LocalRequest.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/LocalRequest.java [hunk 17]
```diff
@@ -1,28 +1,101 @@
 package org.zalando.logbook.jaxrs;
 
+import lombok.AllArgsConstructor;
 import org.zalando.logbook.HttpRequest;
 import org.zalando.logbook.Origin;
 
 import javax.annotation.Nullable;
 import javax.ws.rs.client.ClientRequestContext;
+import java.io.IOException;
```
Comment (whiskeysierra) [inline logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/LocalRequest.java]: Codacy Issue found: [Avoid unused imports such as 'java.io.IOException'](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/RemoteResponse.java [hunk 18]
```diff
@@ -11,16 +12,94 @@
...
+        private final byte[] body;
+
+        @Override
+        public State with() {
+            return new Buffering(body);
+        }
+
+    }
+
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```
Comment (whiskeysierra) [inline logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/RemoteResponse.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-okhttp/src/main/java/org/zalando/logbook/okhttp/LocalRequest.java [hunk 19]
```diff
@@ -13,20 +15,126 @@
...
+
+    }
+
+    private final class Passing extends AbstractState {
 
-    LocalRequest(final Request request) {
-        this.request = request;
     }
 
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```
Comment (whiskeysierra) [inline logbook-okhttp/src/main/java/org/zalando/logbook/okhttp/LocalRequest.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/LocalResponse.java [hunk 20]
```diff
@@ -9,18 +10,89 @@
...
+        @Override
+        public State with() {
+            return new Buffering(stream);
+        }
 
-    public LocalResponse(final ContainerResponseContext context) {
-        this.context = context;
     }
 
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```
Comment (whiskeysierra) [inline logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/LocalResponse.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-httpclient/src/main/java/org/zalando/logbook/httpclient/LocalRequest.java [hunk 21]
```diff
@@ -16,21 +17,110 @@
...
+            return new Buffering(body);
+        }
+
+    }
+
+    private static final class Passing implements State {
+
+    }
+
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```
Comment (whiskeysierra) [inline logbook-httpclient/src/main/java/org/zalando/logbook/httpclient/LocalRequest.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-okhttp2/src/main/java/org/zalando/logbook/okhttp2/LocalRequest.java [hunk 22]
```diff
@@ -13,20 +16,126 @@
...
+
+    }
+
+    private final class Passing extends AbstractState {
 
-    LocalRequest(final Request request) {
-        this.request = request;
     }
 
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```
Comment (whiskeysierra) [inline logbook-okhttp2/src/main/java/org/zalando/logbook/okhttp2/LocalRequest.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

File: logbook-okhttp2/src/main/java/org/zalando/logbook/okhttp2/LocalRequest.java (hunk 13 above)
Comment (whiskeysierra) [inline logbook-okhttp2/src/main/java/org/zalando/logbook/okhttp2/LocalRequest.java]: Codacy Issue found: [Avoid unused imports such as 'org.zalando.fauxpas.FauxPas'](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)

**Comment by radovans** [comment]
Hello, @whiskeysierra, thanks for fix. Do you know when the fix will be realeased? Thank you

**Comment by whiskeysierra** [comment]
:+1:

**Comment by whiskeysierra** [comment]
I hope I can release it tomorrow/today.

**Comment by whiskeysierra** [comment]
@radovans I released 2.0.0-RC.6 to central.

**Comment by radovans** [comment]
Thank you so much.

**Pull request merged**